orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -869,89 +869,217 @@ common::Status HandleSharedInitializer(Graph& graph,
+                              std::vector<std::string>& new_input_names,
+                              std::vector<std::string>& new_output_names,
+                              std::vector<NodeArg*>& send_input_args,
+                              std::vector<NodeArg*>& send_output_args,
+                              std::vector<NodeArg*>& recv_input_args,
+                              std::vector<NodeArg*>& recv_output_args) {
+  std::string cut_index_str = std::to_string(current_stage);
+
+  ORT_RETURN_IF_ERROR(
+    AddNewScalarNodeArgAndInitializer<bool>(graph,
+                                            "send_input_signal" + cut_index_str,
+                                            ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+                                            true, /* initializer data */
+                                            send_input_args,
+                                            new_input_names));
+  ORT_RETURN_IF_ERROR(
+    AddNewScalarNodeArgAndInitializer<size_t>(graph,
+                                              "send_dst_rank" + cut_index_str,
+                                              ONNX_NAMESPACE::TensorProto_DataType_INT64,
+                                              next_stage, /* initializer data */
+                                              send_input_args,
+                                              new_input_names));
+  ORT_RETURN_IF_ERROR(
+    AddNewScalarNodeArgAndInitializer<bool>(graph,
+                                            "recv_input_signal" + cut_index_str,
+                                            ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+                                            true, /* initializer data */
+                                            recv_input_args,
+                                            new_input_names));
+  ORT_RETURN_IF_ERROR(
+    AddNewScalarNodeArgAndInitializer<size_t>(graph,
+                                              "recv_src_rank" + cut_index_str,
+                                              ONNX_NAMESPACE::TensorProto_DataType_INT64,
+                                              current_stage, /* initializer data */
+                                              recv_input_args,
+                                              new_input_names));
+
+  // add output node_arg for send/recv
+  AddNewNodeArg(graph, "send_output_signal" + cut_index_str,
+                ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+                send_output_args, new_output_names);
+
+  AddNewNodeArg(graph, "receive_output_signal" + cut_index_str,
+                ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+                recv_output_args, new_output_names);
+
+  return Status::OK();
+}
+
+common::Status SplitGraphWithMap(Graph& graph,
+                                 std::map<Node*, int>& op_to_stage,
+                                 int nstages,
+                                 std::vector<std::pair<int, int>>& messages,
```

Comment:
>messages [](start = 67, length = 8)

What is it?

Reply:
I updated the documentation of this function. I hope it is clear now. Maybe we don't even need this `messages` vector in this PR. It will be more useful when we consider backward nodes. Should I remove it?